### PR TITLE
[DNM] [LogStreaming Prototype] Add replay command

### PR DESCRIPTION
**Pull request by lopert**
### WHY are these changes introduced?

Part of the LogStreaming Prototype project.

### WHAT is this pull request doing?
- adds a new command `shopify app function replay`, which:
  - reads a folder for run logs
  - presents the list of runs in a selector
  - parses the input from the selected run
  - forwards the run to function-runner

### How to test your changes?

- checkout this branch in your local CLI
- create a `/runs/` folder under your app's extension
- (for now) manually populate it with run logs.
  - ex, for our 10% discount function, you can put this in a json file in that folder
```json
{
  "payload": {
    "input":"{\"cart\":{\"lines\":[{\"quantity\":1,\"merchandise\":{\"__typename\":\"ProductVariant\",\"id\":\"gid:\\\/\\\/shopify\\\/ProductVariant\\\/1\"}}]}}",
    "invocationId":"11111111-ed53-4377-b30f-14e8f4653cfe"
  }
}
```
  - create a few of these to have options in the selector
- run the following command from the CLI directory
```
pnpm run shopify app function replay --path=<path-to-your-ext>
```
ex:
```
pnpm run shopify app function replay --path=/Users/lopert/cli/lopert-cli-local/extensions/10-percent-off
```

### CODE REVIEW

**Comment by lopert** [comment]
Some activity to keep the PR open, will pick this up after ATC work this week.

File: docs-shopify.dev/commands/app-function-replay.doc.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  {
+    title: 'Flags',
+    description: 'The following flags are available for the `app function replay` command:',
+    type: 'appfunctionreplay',
+  },
+  ],
+  category: 'app',
+  related: [
+  ],
+}
```
Comment (DuncanUszkay1) [inline docs-shopify.dev/commands/app-function-replay.doc.ts:32]: Is this going to be publicly visible?

File: packages/app/src/cli/commands/app/function/replay.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import {appFlags} from '../../../flags.js'
+import {selectRunPrompt} from '../../../prompts/dev.js'
+import Command from '@shopify/cli-kit/node/base-command'
+import {globalFlags} from '@shopify/cli-kit/node/cli'
+import {Flags} from '@oclif/core'
+import {writeFile, inTemporaryDirectory} from '@shopify/cli-kit/node/fs'
+import {joinPath} from '@shopify/cli-kit/node/path'
+
+export default class FunctionReplay extends Command {
+  static summary = 'Replays a function locally based on a FunctionRunEvent.'
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/commands/app/function/replay.ts:13]: We've moving away from `Event` and towards `Log`, we should do a sweep here

File: packages/app/src/cli/services/function/replay.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  const runFilePaths = runFileNames.map((runFile) => joinPath(runsFolder, runFile))
+
+  // read contents
+  const runData = await Promise.all(
+    runFilePaths.map((runFile) => {
+      return readFile(runFile)
+    }),
+  )
+
+  // convert promise'd strings to json
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/services/function/replay.ts:24]: IMO the comments here aren't necessary on `main`

File: packages/app/src/cli/services/function/replay.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  const runs = runData.map((run) => JSON.parse(run))
+  return runs
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/services/function/replay.ts:26]: Seems like something we should inline- thoughts?

**Review by DuncanUszkay1 — COMMENTED** [review]
Thoughts:
- How can we merge this while still keeping it hidden? Not to prevent someone from finding out, but just to not mislead them into thinking that it is ready to use
- Let's talk more about the overall spec for the CLI commands on Monday

It seems like the best path forward here is to park this, start finalizing the app dev command including writing the files, and circle back once that's done. We should also have our specs cleared up by then and in every CLI issue.

File: packages/app/src/cli/services/function/replay.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import {joinPath} from '@shopify/cli-kit/node/path'
+import {readFile} from '@shopify/cli-kit/node/fs'
+import {readdirSync} from 'fs'
+
+export async function readFunctionRunsDirectory(functionPath: string) {
```
Comment (andrewhassan) [inline packages/app/src/cli/services/function/replay.ts:5]: Nit: Since this is a function that's exported, I think it'd be good to also type the return type so that the reader doesn't have to grok the implementation.

File: packages/app/src/cli/services/function/replay.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+  // read contents
+  const runData = await Promise.all(
+    runFilePaths.map((runFile) => {
+      return readFile(runFile)
+    }),
+  )
+
+  // convert promise'd strings to json
+  const runs = runData.map((run) => JSON.parse(run))
```
Comment (andrewhassan) [inline packages/app/src/cli/services/function/replay.ts:25]: Instead of reading/parsing every file, since we're using individual files, we could probably get away with saving the files as `<uuid>.json` and returning an array of `{name: "<uuid>", updatedAt: Date}` using the file metadata.

File: packages/app/src/cli/services/function/replay.ts (hunk 5 above)
Comment (lopert) [reply packages/app/src/cli/services/function/replay.ts]: reworked this so this is no longer exported 👍

File: packages/app/src/cli/services/function/replay.ts (hunk 6 above)
Comment (lopert) [reply packages/app/src/cli/services/function/replay.ts]: Hmmm interesting idea. We now have the `identifier` at the end of the file, so we should be able to produce this. I'll give it a whirl.

File: packages/app/src/cli/commands/app/function/replay.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      env: 'SHOPIFY_FLAG_CLIENT_ID',
+      exclusive: ['config'],
+    }),
+    export: Flags.string({
+      char: 'e',
+      hidden: false,
+      description: 'Name of the Web Assembly export to invoke.',
+      default: '_start',
+      env: 'SHOPIFY_FLAG_EXPORT',
+    }),
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/commands/app/function/replay.ts:42]: 💭 Isn't the export being invoked defined by the run itself? E.G. a run using `fetch` couldn't be sensibly rerun using `run`. Perhaps that should be part of the payload so you never have to specify it here?

Probably not urgent for this PR, but would be an improvement for later.

File: packages/app/src/cli/prompts/dev.test.ts [hunk 8]
```diff
@@ -86,6 +125,27 @@ describe('selectOrganization', () => {
   })
 })
 
+describe('selectFunctionRun', () => {
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/prompts/dev.test.ts:128]: I noticed the name of this file is `dev.test.ts`- is our prompt defined in the right place?

File: packages/app/src/cli/commands/app/function/replay.ts (hunk 7 above)
Comment (lopert) [reply packages/app/src/cli/commands/app/function/replay.ts:42]: 🤔 It's one of the flags that we can invoke function runner with so that's how it ended up here. 
The export is defined in the extension's `shopify.extension.toml`, so I think we could maybe improve this by just reading from there?

File: packages/app/src/cli/prompts/dev.test.ts (hunk 8 above)
Comment (lopert) [reply packages/app/src/cli/prompts/dev.test.ts:128]: It is among the other prompts so I think so

File: packages/app/src/cli/commands/app/function/replay.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    const functionPath = flags.path
+
+    const runs = await readFunctionRunsDirectory(functionPath)
+
+    const selectedRun = await selectRunPrompt(runs)
+
+    const input = selectedRun.payload.input
+
+    // dump the output to a file
+    await inTemporaryDirectory(async (tmpDir) => {
```
[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/commands/app/function/replay.ts]: Turns out we can use the [simple input](https://github.com/sindresorhus/execa?tab=readme-ov-file#simple-input) to pass in a string. I've updated the replay code to use this variant of invoking `exec` without touching the existing implementation.

File: packages/app/src/cli/prompts/dev.ts [hunk 10]
```diff
@@ -16,6 +17,23 @@ export async function selectOrganizationPrompt(organizations: Organization[]): P
   return organizations.find((org) => org.id === id)!
 }
 
+export async function selectFunctionRunPrompt(functionRuns: FunctionRunData[]): Promise<FunctionRunData> {
```
[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/prompts/dev.ts]: Thanks for brining this up again. For some reason reading through it again, I understood what you were both asking 🙈 . I've moved this prompt to it's own file under `.../prompts/function/replay/`

File: packages/app/src/cli/services/app-logs/write-app-logs.ts [hunk 11]
```diff
@@ -32,3 +34,12 @@ export const writeAppLogsToFile = async ({
...
   }
 }
+
+function formatTimestampToFilename(timestamp: string): string {
+  // 2024-05-22T15:06:41.827379Z
+  const year = timestamp.substring(0, 10).replace(/-/g, '')
+  const time = timestamp.substring(11, 19).replace(/:/g, '')
+  const microseconds = timestamp.substring(20, 26)
+  const timezone = timestamp.substring(26)
+  return `${year}_${time}_${microseconds}${timezone}`
```
[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/services/app-logs/write-app-logs.ts:44]: I attempted this initially but I couldn't see how to have the `Date` object have microsecond precision. 
```javascript
> date = new Date("2024-06-04T14:34:44.334440Z")
2024-06-04T14:34:44.334Z
> date.toISOString()
'2024-06-04T14:34:44.334Z'
```

File: packages/app/src/cli/services/app-logs/write-app-logs.ts [hunk 12]
```diff
@@ -13,9 +13,11 @@ export const writeAppLogsToFile = async ({
   apiKey: string
   stdout: Writable
 }) => {
-  const fileName = `app_logs_${appLog.log_timestamp}.json`
+  const identifier = randomUUID().substring(0, 6)
+  const formattedTimestamp = formatTimestampToFilename(appLog.log_timestamp)
```
[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/services/app-logs/write-app-logs.ts:17]: Going to lead this as is for now since we're going to rework the response serverside. then we can come back and update this.

File: packages/app/src/cli/services/function/replay.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  const functionRunFileNames = readdirSync(functionRunsDir)
+  const functionRunFilePaths = functionRunFileNames.map((functionRunFile) => joinPath(functionRunsDir, functionRunFile))
+
+  const functionRunData = await Promise.all(
+    functionRunFilePaths.map(async (functionRunFilePath) => {
+      const fileData = await readFile(functionRunFilePath)
+      const parsedData = JSON.parse(fileData)
+      return {
+        ...parsedData,
+        identifier: getIdentifierFromFilename(functionRunFilePath) || 'no identifier',
```
[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/services/function/replay.ts:78]: Oh, because the timestamps contain `_` as well, if I mess with the files and delete the identifier manually, it just takes the the last part of the split, resulting in
```
?  Which function run would you like to replay locally?

>  2024-06-05T15:01:04.154838Z (failure) - 154Z.j
   2024-06-05T15:01:03.420454Z (failure) - 4a0b61
```

[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/services/function/replay.ts]: Since we should never not have an identifier, I've just removed the undefined return with `pop()!`

File: packages/app/src/cli/commands/app/function/replay.ts (hunk 7 above)
Comment (nickwesselman) [reply packages/app/src/cli/commands/app/function/replay.ts]: I agree with @DuncanUszkay1 that this should be part of the payload, the input you are replaying would be specific to a certain target.

Comment (DuncanUszkay1) [reply packages/app/src/cli/commands/app/function/replay.ts]: Will add that to the payload in my upcoming PR that also adds source/source namespace

File: packages/app/src/cli/services/function/replay.ts [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+      }
+    }),
+  )
+
+  return functionRunData
+}
```
[the remark this answers is not in the document]
Comment (lopert) [reply packages/app/src/cli/services/function/replay.ts]: Set a simple limit of 100 for now. We can revisit later if we need to provide an input arg to make this more flexible.

File: packages/app/src/cli/services/function/replay.test.ts [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+  event_type: 'function_run',
+  cursor: '2024-05-31T15:29:47.291530Z',
+  status: 'success',
+  log_timestamp: '2024-05-31T15:29:50.741270Z',
+  identifier: '123456',
+}
+const RUN2_FILENAME = RUN2.log_timestamp.replace(/:/g, '_')
+
+describe('replay', () => {
+  test('reads the app log directory, parses the function runs, and invokes function-runner aa', async () => {
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/services/function/replay.test.ts:59]: Should we add a test for the situation where the app log directory is empty?

File: packages/app/src/cli/services/function/replay.ts [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  const outputAsJson = options.json ? ['--json'] : []
+  const exportName = options.export ? ['--export', options.export] : []
+
+  return exec('npm', ['exec', '--', 'function-runner', '-f', fun.outputPath, ...outputAsJson, ...exportName], {
+    cwd: fun.directory,
+    input,
+    stdout: 'inherit',
+    stderr: 'inherit',
+  })
+}
```
Comment (DuncanUszkay1) [inline packages/app/src/cli/services/function/replay.ts:68]: Leaving a comment here as an anchor for an issue to start using the export in the log itself once that's available. If it's a duplicate feel free to close, but I didn't see another one.

**Review by DuncanUszkay1 — APPROVED** [review]
LGTM aside from dealing with the 'no logs to replay' case explicitly

File: packages/app/src/cli/services/function/replay.test.ts (hunk 15 above)
Comment (lopert) [reply packages/app/src/cli/services/function/replay.test.ts:59]: Good callout! 
I see the other prompts return undefined in this scenario so I'll do that. 
Then here I'll throw an error showing the directory that we tried to find runs in.

File: packages/app/src/cli/services/app-logs/write-app-logs.ts [hunk 17]
```diff
@@ -13,9 +13,12 @@ export const writeAppLogsToFile = async ({
   apiKey: string
   stdout: Writable
 }) => {
-  const fileName = `app_logs_${appLog.log_timestamp}.json`
+  const identifier = randomUUID().substring(0, 6)
+
+  const formattedTimestamp = formatTimestampToFilename(appLog.log_timestamp)
+  const fileName = `${formattedTimestamp}_${identifier}.json`
```
Comment (andrewhassan) [inline packages/app/src/cli/services/app-logs/write-app-logs.ts:19]: Just curious: do we get an error if there's already a file with this name? I'm just wondering in the extreme off chance that there's a collision.

File: packages/app/src/cli/services/function/replay.test.ts [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
```
Comment (andrewhassan) [inline packages/app/src/cli/services/function/replay.test.ts:59]:
```suggestion
  test('reads the app log directory, parses the function runs, and invokes function-runner', async () => {
```

**Review by andrewhassan — APPROVED** [review]
